Sitefinity CLI/Services/PackageService.cs

```diff
@@ -0,0 +1,192 @@
+﻿using Microsoft.Extensions.Logging;
+using Sitefinity_CLI.Model;
+using Sitefinity_CLI.PackageManagement;
+using Sitefinity_CLI.Services.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Sitefinity_CLI.Services
+{
+    internal class PackageService : IPackageService
+    {
+        public PackageService(ISitefinityPackageManager sitefinityPackageManager, IPackageSourceBuilder packageSourceBuilder, IUpgradeConfigGenerator upgradeConfigGenerator, IProjectService projectService)
+        {
+            this.sitefinityPackageManager = sitefinityPackageManager;
+            this.packageSourceBuilder = packageSourceBuilder;
+            this.upgradeConfigGenerator = upgradeConfigGenerator;
+            this.projectService = projectService;
+        }
+
+        public async Task<NuGetPackage> GetLatestCompatibleVersion(string packageId, Version sitefinityVersion, string nugetConfigPath)
+        {
+            IEnumerable<string> versions = await this.sitefinityPackageManager.GetPackageVersions(packageId);
+            IEnumerable<NugetPackageSource> packageSources = await this.GetPackageSources(nugetConfigPath);
+
+            NuGetPackage compatiblePackage = null;
+
+            foreach (string version in versions)
+            {
+                bool isIncompatible = false;
+                NuGetPackage package = await this.sitefinityPackageManager.GetPackageTree(packageId, version, packageSources, package =>
+                {
+                    isIncompatible = this.IsSitefinityPackage(package.Id) && new Version(package.Version) > sitefinityVersion;
+                    return isIncompatible;
+                });
+
+                if (!isIncompatible)
+                {
+                    Version currentVersion = this.GetSitefinityVersionOfDependecies(package);
+                    if (currentVersion <= sitefinityVersion)
+                    {
+                        compatiblePackage = package;
+                        break;
+                    }
+                }
+            }
+
+            return compatiblePackage;
+        }
+
+        public async Task<NuGetPackage> InstallPackage(UpgradeOptions options, IEnumerable<string> sitefinityProjectFilePaths)
```

Comment:
The method is doing more than installing the package. Maybe we should rename and extract some logic.
I see that here we ultimately download the new nuget package in order to be able to build the config used for upggrading sitefinity. We can leave the restore, set target framework and install (this.sitefinityPackageManager.Install) here and name the method something like "PrepareSitefinityUpgradePackage". The output of the method will be a NugetPackage object which we can pass the the UgpradeConfigGenerator